### ISSUE TITLE
Support for passing io_service in to rpc::server?

### ISSUE THREAD

**Issue by nlyan** [rpclib/rpclib#135]
When using the RPC server with other libraries (e.g. Beast, Boost Process or plain old raw ASIO sockets etc), it would be useful to be able to provide your own io_service instance to server class to achieve single-threaded operation.

Currently it's possible to use external ASIO but it's not possible to provide the server class with an io_service instance (preferably via shared_ptr).

**Comment by sztomi** [comment]
Sounds like a good idea. Do you a suggestion for the interface? I'd prefer to not break current code and have this only as an opt-in. Maybe a default param to the constructor?

**Comment by nlyan** [comment]
I've issued a PR as a crude fix. Lightly tested. Had to use a forward declaration of io_service in the server header

**Comment by sztomi** [comment]
Thanks! I'm wondering if `shared_ptr` is really the right call here though. Is it realistic that the original owner of the `io_service` ceases to exist and the server would need to keep it alive? I understand that it's possible, the question is if it's something that is worth paying the price for in the general case.

**Comment by nlyan** [comment]
The io_service has to exist for the duration of all your I/O objects (sockets) etc anyway, so no it's not strictly necessary. Although, it's possible the shared_ptr<io_service> isn't keeping an io_service alive at all. It could be keeping another arbitrary object alive, or nothing alive, thanks to shared_ptrs aliasing constructor.

**Comment by arximboldi** [comment]
I am interested very much in this too! But I do not think that using a `shared_ptr<io_service>` is a good idea. In my opinion, it should just take a reference.

**Comment by sytelus** [comment]
In my view this violates the abstraction principle for the rpclib. A good design for rpclib should not leak internal implementation details in interface. Once you advertise io_service through interface you have confined yourself not be able to change your internal implementation to something other than Boost. One way to avoid this is to put methods users in io_service as abstract interface and expose that interface outside. Then one can implement this abstract interface and supply it to rpclib.

**Comment by sztomi** [comment]
Good point @sytelus , I'm a bit on the fence regarding this for that exact reason. My thinking is that such function could be exposed via a macro for users that want to rely on an implementation detail like this (i.e. add a define to be able to use that function). On the other hand I can already see people complaining when rpclib drops asio in the future (no concrete plans for that at this moment!)
